src/client/builtin_scene/web_content_renderer.cpp

```diff
@@ -804,6 +627,360 @@ namespace builtin_scene::web_renderer
+      float borderLeft = fragment.border().left();
+
+      float paddingTop = fragment.padding().top();
+      float paddingRight = fragment.padding().right();
+      float paddingBottom = fragment.padding().bottom();
+      float paddingLeft = fragment.padding().left();
+
+      SkRect contentRect = SkRect::MakeLTRB(
+        borderBox.fLeft + borderLeft + paddingLeft,
+        borderBox.fTop + borderTop + paddingTop,
+        borderBox.fRight - borderRight - paddingRight,
+        borderBox.fBottom - borderBottom - paddingBottom);
+
+      // Adjust radii for the content box
+      SkVector radii[4];
+      for (int i = 0; i < 4; i++)
+      {
+        SkVector originalRadius = roundedRect.radii(static_cast<SkRRect::Corner>(i));
+        // Reduce radii by border and padding width (but don't go below 0)
+        float totalReduction = (i == 0 || i == 3) ? (borderLeft + paddingLeft) : (borderRight + paddingRight); // Simplified
+        radii[i] = SkVector{std::max(0.0f, originalRadius.x() - totalReduction),
+                            std::max(0.0f, originalRadius.y() - totalReduction)};
+      }
+      clippingArea.setRectRadii(contentRect, radii);
+    }
+
+    // For border-box (default) and text, return the original rounded rect
+    return clippingArea;
+  }
+
+  optional<SkPath> RenderBackgroundSystem::createTextPath(const std::string &textContent,
+                                                          const WebContent &content)
+  {
+    if (textContent.empty())
+      return nullopt;
+
+    SkPath textPath;
+
+    // Create paragraph to get better text bounds
+    auto clientContext = TrClientContextPerProcess::Get();
+    auto fontCollection = clientContext->getFontCacheManager();
+    auto paragraphStyle = content.paragraphStyle();
+    auto paragraphBuilder = ParagraphBuilder::make(paragraphStyle, fontCollection);
+    paragraphBuilder->pushStyle(paragraphStyle.getTextStyle());
+    paragraphBuilder->addText(textContent.c_str(), textContent.size());
+    paragraphBuilder->pop();
+
+    auto layoutWidth = round(content.fragment()->contentWidth()) + 1.0f;
+    auto paragraph = paragraphBuilder->Build();
+    paragraph->layout(layoutWidth);
+
+    // Try to get more accurate text bounds from the paragraph
+    const auto &fragment = content.fragment();
+    if (!fragment.has_value())
+      return textPath;
+
+    // Get actual text metrics from the paragraph
+    float textHeight = paragraph->getHeight();
+    float textWidth = paragraph->getMaxIntrinsicWidth();
+
+    // Use the actual text dimensions but limit to content area
+    float actualWidth = std::min(textWidth, fragment->contentWidth());
+    float actualHeight = std::min(textHeight, fragment->contentHeight());
+
+    // Create a more accurate rectangular approximation of text bounds
+    // This is still a simplified approach, but better than using full content area
+    textPath.addRect(SkRect::MakeWH(actualWidth, actualHeight));
```

Comment:
You need to create the text path with its font shapes, not a rectangle that the text placed at.